### PR TITLE
[SPARK-18999][SQL][minor] simplify Literal codegen

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`Literal` can use `CodegenContex.addReferenceObj` to implement codegen, instead of `CodegenFallback`.  This can also simplify the generated code a little bit, before we will generate: `((Expression) references[1]).eval(null)`, now it's just `references[1]`.

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -92,10 +92,10 @@ class CodegenContext {
    * This is for minor objects not to store the object into field but refer it from the references
    * field at the time of use because number of fields in class is limited so we should reduce it.
    */
-  def addReferenceObj(obj: Any): String = {
+  def addMinorReferenceObj(obj: Any, className: String = null): String = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: To be consistent with the above comments, how about `addReferenceMinorObj`?

**Comment by gatorsmile** [comment]
LGTM, except one minor comment.

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by kayousterhout** [comment]
This commit introduced a bug where IN doesn't work right for Infinity / -Infinity (JIRA [here](https://issues.apache.org/jira/browse/SPARK-19072)).  I'm not sure how to fix the underlying bug (or if this PR should just be reverted) -- @cloud-fan @gatorsmile can one of you fix this?  

The relevant test also failed for this PR the first time tests were run -- remember to make sure that test failures aren't related to the PR!

**Comment by cloud-fan** [comment]
Sorry it's my bad, I should take a look at the test result before retest it. I've sent a PR to fix it.

**Pull request closed without merging**